### PR TITLE
Simply multichain running

### LINKED ISSUE #1636: Simplify multichain running
### Description

Currently running a multichain project involves managing each chain instance separately, having a manifest file for each chain. There is lots of room for improvement to simplify this process

### Goals

To make it easier to run multichain projects (multiple nodes) both locally and on the managed service.
- Developers can codegen their project by running a single `yarn codegen` command that defaults to the Parent manifest
- Developers can build all child chain projects by running a single `yarn build` command that defaults to the Parent manifest
- Developers can build a single child chain by specifying the individual manifest on build
- Make generating the Docker Compose easy, one idea is a cli command to create a new chain manifest, add it to the Parent Manifest, and add it to the Docker Compose??

Then work with hosted service to
- Developers can publish a SubQuery multichain project (including all child chains) using the subql publish command (assuming their parent manifest is setup correctly

### Parent Manifest

A parent manifest seems like the obvious solution. This will include references to each chains specific project manifest as well as query service options that will override what is specified in each chain specific project.

Example:
```yaml
# subquery-multichain.yaml
specVersion: 1.0.0

projects:
   - ./path/to/chain1-project.yaml
   - ./path/to/chain2-project.yaml
   - ...

query:
  name: "@subql/query"
  version: "*"


```

### Stages

1. Define in code the manifest spec as we have with  current manifests.
2. CLI to publish and create multichain files. 
3. Further tooling
    * Validate schema matches for all projects
    * Run the multichain project locally

### PR DESCRIPTION
**Pull request by DeveloperInProgress**
# Description
Allow multichain projects to be built with single codegen and build command. Add new child manifests with cli. 

Fixes #1636 

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Checklist

- [ ] I have tested locally
- [ ] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [ ] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My code is up to date with the base branch
- [ ] I have updated relevant changelogs. [We suggest using chan](https://github.com/geut/chan/tree/main/packages/chan)

### CODE REVIEW

File: packages/cli/src/commands/add-chain.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import {Command, Flags} from '@oclif/core';
+import {addChain} from '../controller/add-chain-controller';
+
+export default class AddChain extends Command {
+  static description = 'Add new chain manifest to multi-chain configuration';
+
+  static flags = {
+    multichain: Flags.string({char: 'm', description: 'specify multichain manifest file path'}),
```
Comment (jiqiang90) [inline packages/cli/src/commands/add-chain.ts:11]: little confused here with `multichain manifest` and `new chain manifest`

File: packages/cli/src/controller/add-chain-controller.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+  const defaultMultichainManifestPath = path.resolve(process.cwd(), 'multichain-manifest.yaml');
+
+  return multichain ? path.resolve(multichain) : defaultMultichainManifestPath;
+}
+
+function loadOrCreateMultichainManifest(multichainManifestPath: string): ProjectManifestParentV1_0_0 {
+  let multichainManifest: ProjectManifestParentV1_0_0;
+
+  if (fs.existsSync(multichainManifestPath)) {
+    multichainManifest = yaml.load(fs.readFileSync(multichainManifestPath, 'utf8')) as ProjectManifestParentV1_0_0;
```
Comment (stwiname) [inline packages/cli/src/controller/add-chain-controller.ts:34]: I think we had a new way of loading and writing yaml that retained the comments in the file

File: packages/cli/src/controller/add-chain-controller.ts [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+function loadDockerComposeFile(dockerComposePath: string): DockerComposeType | undefined {
+  if (fs.existsSync(dockerComposePath)) {
+    return yaml.load(fs.readFileSync(dockerComposePath, 'utf8')) as DockerComposeType;
+  }
+}
+
+function getSubqlNodeService(dockerCompose: DockerComposeType): any {
+  const subqlNodeServices = Object.keys(dockerCompose.services).filter((key) =>
```
Comment (stwiname) [inline packages/cli/src/controller/add-chain-controller.ts:115]: You can use `Object.values` and find if you only need the one.

File: packages/cli/src/controller/add-chain-controller.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  multichainManifest.projects.push(chainManifestPath);
+}
+
+function loadDockerComposeFile(dockerComposePath: string): DockerComposeType | undefined {
+  if (fs.existsSync(dockerComposePath)) {
+    return yaml.load(fs.readFileSync(dockerComposePath, 'utf8')) as DockerComposeType;
+  }
+}
+
+function getSubqlNodeService(dockerCompose: DockerComposeType): any {
```
Comment (stwiname) [inline packages/cli/src/controller/add-chain-controller.ts:114]: Return type please

File: packages/cli/src/commands/publish.ts [hunk 5]
```diff
@@ -48,14 +48,16 @@ export default class Publish extends Command {
...
     }
-    const cid = await uploadToIpfs(project.manifest, authToken.trim(), flags.ipfs).catch((e) => this.error(e));
-    await createIPFSFile(location, cid);
-
-    if (!flags.output) {
-      this.log('Uploading SubQuery project to IPFS');
-      this.log(`SubQuery Project uploaded to IPFS: ${cid}`);
-    } else {
-      this.log(`${cid}`);
+    for (const manifest of project.manifests) {
```
Comment (stwiname) [inline packages/cli/src/commands/publish.ts:51]: If its multichain, we should create a directory on ipfs

**Review by stwiname — COMMENTED** [review]
This is a good opportunity to have tests. Everything here can be easily testable.

There can also be many improvements to types. We should not use `any` except when we cannot find any alternative.

Some of these changes don't look backwards compatible. The user should still be able to use the CLI as is now without having to have a multichain manifest

**Review by stwiname — CHANGES_REQUESTED** [review]
I did some testing and all the commands i tried failed

## Issues from testing:
### Non multichain projects:
* `./bin/run codegen -f ~/Projects/subql-dictionary/moonbeam-evm` and ``./bin/run codegen -f ~/Projects/subql-dictionary/moonbeam-evm/project.yaml` both fail with ` Error: Could not find project manifest under dir .`
* `./bin/run publish -f ~/Projects/subql-dictionary/moonbeam-evm/project.yaml` and `./bin/run publish -f ~/Projects/subql-dictionary/moonbeam-evm` fails with `Error: unknown location: project.yaml`
### Multichain project:
* codegen is the same as above
* Publish cannot find project file: `Error: unknown location: moonbeam_project.yaml`
* `./bin/run multi-chain:add -f ~/Projects/api-nft-erc721/subquery-multichain.yaml` doesnt modify docker compose in the right place `Error: Docker Compose file does not exist at the specified location: /Users/scotttwiname/Projects/subql-main/packages/cli/docker-compose.yml`

File: packages/cli/src/commands/multi-chain/add.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import {Command, Flags} from '@oclif/core';
+import {addChain} from '../../controller/add-chain-controller';
+
+export default class MultiChainAdd extends Command {
+  static description = 'Add new chain manifest to multi-chain configuration';
+
+  static flags = {
+    multichain: Flags.string({char: 'm', description: 'specify multichain manifest file path'}),
```
Comment (stwiname) [inline packages/cli/src/commands/multi-chain/add.ts:11]: I think `-f` should be the multichain manifest file.

File: packages/cli/src/commands/multi-chain/add.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2020-2022 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import {Command, Flags} from '@oclif/core';
+import {addChain} from '../../controller/add-chain-controller';
+
+export default class MultiChainAdd extends Command {
```
Comment (stwiname) [inline packages/cli/src/commands/multi-chain/add.ts:7]: Can we make this interactive as well please

**Review by stwiname — CHANGES_REQUESTED** [review]
After more testing I'm still facing issues:

* A directory with no `project.yaml` exits without any error message for `codegen` and `publish`
* A directory with a `subquery-multichain.yaml` file exits without doing anything for `codegen` and `publish`
* Publishing a multichain project with one chain only logs one CID and its the CID of the one chain. It doesn't say which chain it is for or if its the multi-chain manifest
* Adding a chain to multichain doesn't log anything. I don't know if it works without viewing the files
* Adding the same chain to multichain adds it again to the manifest file. It should ignore it if its already there.
* Docker compose doesn't include the `multi-chain` flag

**Comment by DeveloperInProgress** [comment]
> * A directory with a `subquery-multichain.yaml` file exits without doing anything for `codegen` and `publish`

right now the default name for a multichain manifest file is `multichain-manifest.yaml`

**Comment by stwiname** [comment]
> > * A directory with a `subquery-multichain.yaml` file exits without doing anything for `codegen` and `publish`
> 
> right now the default name for a multichain manifest file is `multichain-manifest.yaml`

Can it be `subquery-multichain` like in the issue please

File: packages/cli/src/commands/multi-chain/add.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  static description = 'Add new chain manifest to multi-chain configuration';
+
+  static flags = {
+    multichain: Flags.string({char: 'f', description: 'specify multichain manifest file path'}),
+    chainManifestPath: Flags.string({char: 'c', description: 'path to the new chain manifest'}),
+    chainId: Flags.string({description: 'ID of the new chain'}),
+    schema: Flags.string({char: 's', description: 'specify schema path for the new chain manifest'}),
```
Comment (stwiname) [inline packages/cli/src/commands/multi-chain/add.ts:15]: How come a schema needs to be provided? They should all have the same schema

**Review by stwiname — CHANGES_REQUESTED** [review]
It seems codegen is being run for the same files multiple times:
```
➜ ./bin/run codegen -f  ~/Projects/api-nft-erc721
===============================
---------Subql Codegen---------
===============================
Clearing generated files
Cleared generated directory: /Users/scotttwiname/Projects/api-nft-erc721/src/types
* Schema Nft generated !
* Schema Collection generated !
* Schema Transfers generated !
* Schema Address generated !
* Schema Network generated !
* Models index generated !
* Abi Interface Erc721 generated
* Abi Interface Erc1155 generated
* Types index generated !
* Abi Interface Erc721 generated
* Abi Interface Erc1155 generated
* Types index generated !
* Abi Interface Erc721 generated
* Abi Interface Erc1155 generated
* Types index generated !
```

Adding a new chain with chain id:
```
./bin/run multi-chain:add -f ~/Projects/api-nft-erc721/subquery-multichain.yaml  --chainId=592
Enter the path to the new chain manifest: astar.yaml
    TypeError: The "path" argument must be of type string or an instance of Buffer or URL. Received an instance of YAMLSeq
    Code: ERR_INVALID_ARG_TYPE
```
If the current directory is a project `-f` should be unnecessary

Please do more testing. I stopped once I found enough issues

**Review by stwiname — COMMENTED** [review]
I ended up wiith a docker compose that had this section:
```
    command:
      - -f=app/project-astar.yaml
      - --disable-historical
      - -f=/app/project-moonbeam.yaml
      - --db-schema=app
      - --batch-size=30
```

The logging for adding a chain seems excessive:
```
Validating chain manifest...
Chain manifest: project-astar.yaml
Chain manifest schema path: /Users/scotttwiname/Projects/api-nft-erc721/schema.graphql
Validating project: project-moonbeam.yaml
Project schema path: /Users/scotttwiname/Projects/api-nft-erc721/schema.graphql
Adding chain manifest path: project-astar.yaml
Updating Docker Compose for chain manifest: /Users/scotttwiname/Projects/api-nft-erc721/project-astar.yaml
Service name: subquery-node-project-astar
Docker Compose file path: /Users/scotttwiname/Projects/api-nft-erc721/docker-compose.yml
Created new service configuration: subquery-node-project-astar
Docker Compose file updated successfully at: /Users/scotttwiname/Projects/api-nft-erc721/docker-compose.yml
```

How can I see the directory IPFS CID when uploading a multichain project?

Theres a couple of types improvements to be made, see the github annotations

File: packages/cli/src/controller/add-chain-controller.ts [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+  console.log(`Created new service configuration: ${serviceName}`);
+
+  (services as YAMLMap).add({key: serviceName, value: subqlNodeService});
+
+  // Save the updated Docker Compose file
+  fs.writeFileSync(dockerComposePath, dockerCompose.toString());
+  console.log(`Docker Compose file updated successfully at: ${dockerComposePath}`);
+}
+
+function generateChainManifest(chainId: string): string {
```
[the remark this answers is not in the document]
Comment (DeveloperInProgress) [reply packages/cli/src/controller/add-chain-controller.ts:244]: It was expected in the issue. but, currently the implementation is unclear

File: packages/cli/src/commands/publish.ts [hunk 10]
```diff
@@ -48,14 +48,35 @@ export default class Publish extends Command {
...
-      this.log('Uploading SubQuery project to IPFS');
-      this.log(`SubQuery Project uploaded to IPFS: ${cid}`);
-    } else {
-      this.log(`${cid}`);
+    let directory: string;
+    if (project.manifests.length > 1) {
+      directory = path.basename(project.root); // Using project root name as directory
+      // or use a timestamp for uniqueness
+      directory += `_${new Date().getTime()}`;
     }
```
Comment (stwiname) [inline packages/cli/src/commands/publish.ts:56]: I don't think we need this directory stuff. Its not relevant and just makes the logging confusing

File: packages/cli/src/controller/add-chain-controller.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (stwiname) [reply packages/cli/src/controller/add-chain-controller.ts:262]: Lets remove this code as discussed

File: packages/cli/src/controller/codegen-controller.ts [hunk 11]
```diff
@@ -513,18 +511,21 @@ export async function generateModels(projectPath: string, schema: string): Promi
...
-  specVersion: string,
-  templates: TemplateKind[]
-): Promise<void> {
+export async function generateDatasourceTemplates(projectPath: string, templates: TemplateKind[]): Promise<void> {
   const props = templates.map((t) => ({
     name: t.name,
     args: hasParameters(t) ? 'Record<string, unknown>' : undefined,
   }));
+
+  const propsWithoutDuplicates = props.filter((prop, index) => {
```
Comment (stwiname) [inline packages/cli/src/controller/codegen-controller.ts:520]: https://lodash.com/docs/4.17.15#uniqBy

**Review by stwiname — COMMENTED** [review]
- Publishing doesn't upload `subquery-multichain.yaml` in the directory
- There's still a few linter warnings (see annotations)

**Review by stwiname — COMMENTED** [review]
When uploading subquery-multichain we need to replace the file paths with ipfs links to the chain files

**Review by stwiname — APPROVED** [review]

**Pull request merged**